benchmark/python/benchmark_e2e.py

```diff
@@ -0,0 +1,184 @@
+    import pandas as pd
+    df = pd.DataFrame(
+        results,
+        columns=[
+            "Batch Size",
+            "Prompt Length",
+            "Tokenization Throughput (tps)",
+            "Tokenization Latency (ms)",
+            "Prompt Processing Throughput (tps)",
+            "Prompt Processing Latency (ms)",
+            "Token Generation Throughput (tps)",
+            "Token Generation Latency (ms)",
+            "Sampling Throughput (tps)",
+            "Sampling Latency (ms)",
+        ],
+    )
+    df = df.transpose()  # This line swaps the rows and columns
+    df.to_csv(filename, header=False)
+    print(f"Results saved in {filename}!")
+
+def main(args):
+    # Get user arguments
+    num_repititions = args.repetitions
+    generation_length = args.generation_length
+    batch_size, prompt_length = args.batch_size, args.prompt_length
+
+    # Get tokenizer, and model
+    if args.verbose: print(f"Loading model... ")
+    model=og.Model(f'{args.input_folder}', og.DeviceType.CPU if args.device == 'cpu' else og.DeviceType.CUDA)
+    if args.verbose: print("Model loaded")
+    tokenizer = model.create_tokenizer()
+
+    # Generate prompt
+    prompt = [generate_prompt(model, tokenizer, prompt_length)] * batch_size
+    if args.verbose: print("Running warmup runs...")
+    for i in range(args.warmup):
+        if args.verbose: print(f"Running warmup repetition {i+1}...")
+        tokens = tokenizer.encode_batch(prompt)
+        params = og.GeneratorParams(model)
+        params.max_length = prompt_length + generation_length
+        params.input_ids = tokens
+        generator = og.Generator(model, params)
+        while not generator.is_done():
+            generator.compute_logits()
+            generator.generate_next_token_top_k_top_p(args.top_k, args.top_p, 1.0)
+        if args.print_model_output: print(tokenizer.decode(generator.get_sequence(0).get_array()))
+
+    tokenize_times = []
+    prompt_times = []
+    token_gen_times = []
+    sampling_times = []
+    if args.verbose: print(f"Running benchmark for batch size = {batch_size}, prompt length = {prompt_length}")
+    for i in range(num_repititions):
+        if args.verbose: print(f"Running repetition {i+1}...")
+        # Prepare run
+        max_length = prompt_length + generation_length
+        params = og.GeneratorParams(model)
+        params.max_length = max_length
+        params.input_ids = tokens
+        generator = og.Generator(model, params)
+
+        # Measure tokenization
+        tokenize_start_time = time.time()
+        tokens = tokenizer.encode_batch(prompt)
+        tokenize_end_time = time.time()
+        tokenize_times.append(tokenize_end_time - tokenize_start_time)
+
+        # Measure prompt processing
+        prompt_start_time = time.time()
+        generator.compute_logits()
+        prompt_end_time = time.time()
+        prompt_times.append(prompt_end_time - prompt_start_time)
+
+        sampling_start_time = time.time()
```

Comment:
For some measurements such as sampling where the time difference is very small, should we use [`time.perf_counter()`](https://docs.python.org/3/library/time.html#time.perf_counter) instead of `time.time()`?

Reply:
should we use it everywhere or just in the very small ones like sampling and tokenization?

Reply:
We can use it everywhere.